designs/2024-baseline-support/README.md

```diff
@@ -33,44 +33,54 @@ This can be counterintuitive for enabling new rules as `error`, since the develo
+Here is what the baseline file looks like. This indicates that the file `"src/app/components/foobar/foobar.component.ts"` has one error for the rule `@typescript-eslint/no-explicit-any` that is acceptable to be ignored.
+
+All paths are relative to CWD.
 
 ```
 {
-  "/home/user/project/src/app/components/foobar/foobar.component.ts": {
+  "src/app/components/foobar/foobar.component.ts": {
     "@typescript-eslint/no-explicit-any": 1
   }
 }
 ```
 
 ### Generating the baseline
 
-A new option `--baseline` can be added to ESLint CLI. When provided, the baseline is generated and saved in `.eslint-baseline.json`. If the file already exists, it gets over-written. Note that this is a boolean flag option (no values are accepted). For example:
+A new option `--baseline` can be added to ESLint CLI. When provided, the baseline is generated and saved in `.eslintbaseline`. If the file already exists, it gets over-written. Note that this is a boolean flag option (no values are accepted). For example:
 
 ``` bash
 eslint --baseline ./src
 ```
 
-The above will go through each result item and messages, and count the number of errors (`severity == 2`). If one or more such messages are found, the necessary details must be stored in the baseline file. The process should take place right after the fixes are applied to avoid counting errors that are about to be fixed.
+The above goes through each result item and messages, and counts the number of errors (`severity == 2`). If one or more such messages are found, the necessary details are stored in the baseline file. Note that the file is a relative path to CWD.
 
-By default, the baseline file is saved at `.eslint-baseline.json` . To control where the baseline is saved, another option can be introduced  `--baseline-location`. That is a string argument specifying the file or a directory. If a directory is specified, a baseline file is created inside the specified folder.
+By default, the baseline file is saved at `.eslintbaseline` . To control where the baseline is saved, another option can be introduced  `--baseline-location`. That is a string argument specifying where the baseline must be saved.
 
 ``` bash
 eslint --baseline --baseline-location /home/user/project/mycache
 ```
 
+To implement this, we will need to add the two new options in `default-cli-options.js`, adjust the config for optionator and add the two new options as comments and arguments for both eslint and cli-engine. Documentation must be updated as well to explain the newly introduced options.
+
+On top of that, we will need to adjust `cli.js` to check if `--baseline` was provided and set to true, right after the fixes are done and the warnings are filtered out, to avoid counting errors that are about to be fixed. A new method `generateBaseline` can be introduced in both `eslint.js` and `eslint-legacy.js` - the method must be called only and only if `--baseline` was provided and set to true.
```

Comment:
Is there a reason `generateBaseline()` would need to be in `eslint.js`? (Note: We aren't making any further changes to `eslint-legacy.js`)

It seems like the functionality to generate a baseline could be encapsulated in a single function that takes a result set and returns the baseline format?

Thinking this through, it might actually be better to create a `ResultBaseline` (or some other rational name) class that can both generate a baseline and validate results against a baseline.

What do you think?

Reply:
The reason this was suggested was mostly to be aligned with fix and others i.e. `outputFixes` . I like the suggested approach though, this should facilitate testing too.  I am outlining below a possible implementation:

``` js
/**
 * @typedef {Record<string, { count: number }>} BaselineResult
 */

class BaselineResultManager {
    /**
     * Creates a new instance of BaselineResultManager.
     * @param {string} baselineLocation The location of the baseline file.
     */
    constructor(baselineLocation) {}

    /**
     * Generates the baseline from the provided lint results.
     * @param {LintResult[]} results The lint results.
     * @returns BaselineResult[]
     */
    generateBaseline(results) 

    /**
     * Checks the provided baseline results against the lint results.
     * It filters and returns the lint results that are not in the baseline.
     * It also returns the unmatched baseline results.
     * 
     * @param {LintResult[]} results The lint results.
     * @param {BaselineResult[]} baseline The baseline.
     * @return {
     *   results: LintResult[],
     *   unmatched: BaselineResult[]
     * }
     */
    applyBaseline(results, baseline)
    
    /**
     * Loads the baseline from the file.
     * @returns BaselineResult[]
     */
    loadBaseline()

    /**
     * Saves the baseline to the file.
     * @param {BaselineResult[]}
     * @returns void
     * @private
     */
    saveBaseline(baseline)
}
```

The resolution of the baseline location should happen outside of the above class. An idea is to make `getCacheFile` a bit more abstract so that we could inject the prefix i.e. `.cache_` or `.baseline` when a directory is provided. This way both `cache-location` and `baseline-location` are consistent and following the same pattern.

Levaraging the above the changes in `cli.js` should look something like:

``` js
        const baselineFileLocation = getCacheFile(baseline, cwd, '_baseline');
        if (options.baseline || fs.existsSync( baseline )) {
            const baselineManager = new BaselineResultManager(baselineFileLocation);
            let loadedBaselineRecords = [];
            if (options.baseline) {
                loadedBaselineRecords = baselineManager.generateBaseline(resultsToPrint);
            } else {
                loadedBaselineRecords = baselineManager.loadBaseline();
            }

            const baselineResults = await baselineManager.applyBaseline(resultsToPrint, loadedBaselineRecords);

            if (baselineResults.unmatched.length > 0) {
                // exit with errors
            }

            resultsToPrint = baselineResults.results;
        }
```


Many thanks for your feedback btw!! Let me know I should include the above examples in the RFC.

Reply:
Yeah, I think that's a good starting point. We can iterate more on the API once it's included.

Reply:
More implementation details based on the above are now included in the PR.